lib/timeline/frame_flame_chart.dart

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../ui/elements.dart';
+import '../ui/primer.dart';
+import 'timeline_protocol.dart';
+
+class FrameFlameChart extends CoreElement {
+  FrameFlameChart() : super('div') {
+    layoutVertical();
+    flex();
+
+    // TODO(devoncarew): listen to tab changes
+    content = div(c: 'frame-timeline')..flex();
+
+    final PTabNav tabNav = PTabNav(<PTabNavTab>[
+      PTabNavTab('Frame timeline'),
+      PTabNavTab('Widget build info'),
+      PTabNavTab('Skia picture'),
+    ]);
+
+    add(<CoreElement>[
+      tabNav,
+      content,
+    ]);
+
+    content.element.style.whiteSpace = 'pre';
+    content.element.style.overflow = 'scroll';
+  }
+
+  TimelineFrameData data;
+  CoreElement content;
+
+  void updateData(TimelineFrameData data) {
+    this.data = data;
+
+    content.clear();
+
+//    if (data != null) {
+//      StringBuffer buf = new StringBuffer();
+//
+//      for (TimelineThread thread in data.threads) {
+//        buf.writeln('${thread.name}:');
+//
+//        for (TEvent event in data.events) {
+//          if (event.threadId == thread.threadId) {
+//            event.format(buf, '  ');
+//          }
+//        }
+//      }
+//
+//      content.text = buf.toString();
+//    }
+
+    if (data != null) {
+      _render(data);
+    }
+  }
+
+  void _render(TimelineFrameData data) {
+    const int leftIndent = 130;
+    const int rowHeight = 25;
+
+    const double microsPerFrame = 1000 * 1000 / 60.0;
+    const double pxPerMicro = microsPerFrame / 1000.0;
+
+    int row = 0;
+
+    final int microsAdjust = data.frame.startMicros;
+
+    int maxRow = 0;
+
+    Function drawRecursively;
+
+    drawRecursively = (TimelineThreadEvent event, int row) {
```

Comment:
Make this a named function instead of a closure
eg.
void drawRecursively(TimelineThreadEvent event, int row) {

}
That will make debuggers do a bit of a better job as is more idiomatic dart.

Reply:
done.